lang: core: util: add a new func for URL parsing

lourencovales opened the pull request:
This is a first attempt to adding a new function for URL parsing, using go's net/url package and the simple API. This is still a barebones implementation, there's possibility to expose more information that the stdlib exposes.

Related: https://github.com/purpleidea/mgmt/issues/606

purpleidea commented:
I expect the basic tests will fail. See if you can figure out why by reading the logs, and ping when a fix is in. Looks like you're on the right track, but no hints =D This is as much about learning how CI works and so on.

purpleidea commented:
This will still fail for the same reason.

purpleidea commented:
So your commit message will fail the tests, and secondly, since this is pre-merge, we can also hide any mistakes. So you will need to rebase both of these commits into a single one.

Remember to look at a `git diff` right before you push to see if every single byte you changed is what you wanted. HTH

purpleidea commented:
Lastly, for this patch consider adding a short test in lang/interpret_test/TestAstFunc2/ folder.

lourencovales commented:
Oops, I missed the part where you were asking for tests all in one commit. In any case, I fixed the basic tests, and the previous commit I had also fixed the function signature, which was also causing trouble with the tests. I'll work on tests now. Do you also want an all-in-one commit in this case?

purpleidea commented:
> Do you also want an all-in-one commit in this case?

Yeah it's probably good to have everything here in one commit. A single .mcl test will probably be good enough!

purpleidea commented on lang/core/util/url_parser.go:
```diff
@@ -0,0 +1,81 @@
+	"github.com/purpleidea/mgmt/util/errwrap"
+)
+
+const structComp = "struct{Scheme str; Host str; Path str; Query str}"
```
I think the field names should be lowercase. Let's name this var urlParserReturnType too.

lourencovales replied:
I had the field names with lowercase, but then the test I built complained that the struct had an exported field:

`interpret_test.go:802: test #226: funcs: Erroring func printf@0xc000980000: panic in Stream of func printf@0xc000980000: struct has unexported field: scheme`

Ack'd on the variable name.

purpleidea commented on lang/core/util/url_parser.go:
```diff
@@ -0,0 +1,81 @@
+// given URL doesn't contain a scheme, as this is a limitation of the underlying
+// net/url library.
+func URLParser(ctx context.Context, input []types.Value) (types.Value, error) {
+	i := input[0].Str()
```
Nit: Short letter vars have an unwritten convention: use `s` instead here. `i` is commonly for integers or counters.
OR since we don't need the var elsewhere, substitute this into the url.Parse call directly.

lourencovales replied:
fair on both points; in this situation it does make more sense to avoid declaring a new variable.

purpleidea commented on lang/core/util/url_parser.go:
```diff
@@ -0,0 +1,81 @@
+	if err := v.Set("Path", &types.StrValue{V: u.Path}); err != nil {
+		return nil, errwrap.Wrapf(err, "invalid path value")
+	}
+	if err := v.Set("Query", &types.StrValue{V: u.RawQuery}); err != nil {
```
should this be `query` or `raw_query` ?

lourencovales replied:
The URL type on the net/url package does not export a "query" field: https://pkg.go.dev/net/url#URL
They do have a Query function, but this would make a logic here a tad more complex - still doable, but up to you.

purpleidea replied:
My comment is about how to name this in the struct... What do we want the user to see?

purpleidea commented on lang/core/util/url_parser.go:
```diff
@@ -0,0 +1,81 @@
+	}
+
+	v := types.NewStruct(types.NewType(structComp))
+	if err := v.Set("Scheme", &types.StrValue{V: u.Scheme}); err != nil {
```
optional: You _could_ store the field names as a constant, like urlParserFieldScheme and so on... The value of this is that they can be used in the type above and it enforces a logical connection that the two are the same "string".

Know what I mean?

lourencovales replied:
not sure I understand here. Are you suggestion doing something like this?

```
const urlParserFieldScheme = "Scheme"
const urlParserFieldHost = "Host"
...
const urlParserReturnType = "struct{" + urlParserFieldScheme + " str; " + urlParserFieldHost + " str; ...
...
if err := v.Set(urlParserFieldScheme, &types.StrValue{V: u.Scheme}); err != nil {
...
```

purpleidea replied:
Sort of yes. Hint: If you followed this practice, you would likely not have hit a bug!
(Although I'd recommend fmt.Sprintf instead of + to make this readable.)

purpleidea commented on lang/core/util/url_parser.go:
```diff
@@ -0,0 +1,81 @@
+	})
+}
+
+// URLParser takes an URL as a string, and finds the different components of
+// said URL - scheme, host, path, and query. The function will error out if the
+// given URL doesn't contain a scheme, as this is a limitation of the underlying
+// net/url library.
```
```suggestion
// URLParser takes a URL as a string, and extracts the different components of
// the URL: scheme, host, path, and query. The function will error out if the
// given URL doesn't contain a valid URL, as this is a requirement of the underlying
// net/url library. Note that additional fields may be added to the output in the future.
```

(How about this? Make sure to wrap it correctly.)

lourencovales replied:
This does read better, appreciate the corrections!

purpleidea commented on lang/interpret_test/TestAstFunc2/url_parser.txtar:
```diff
@@ -0,0 +1,19 @@
+test [fmt.printf("%v", util.url_parser($url1)),] {}
+test [fmt.printf("%v", util.url_parser($url2)),] {}
+test [fmt.printf("%v", util.url_parser($url3)),] {}
+
```
nit: You could also do:

```
$whatever = util.url_parser($url1)
test [$whatever->scheme,] {}
```

for example...

lourencovales replied:
What would be the advantage here? Not sure I grasp the implications.

purpleidea replied:
Functions in mcl can only return one value (which is the case for most programming languages). Golang happens to allow multiple btw, neat! In any case, we return a struct. But the user _probably_ wants to access just one or two values from that struct. So prove that they can get them out =D

purpleidea reviewed (commented):
Patch is excellent, thank you!

I've nitpicked this patch like crazy since we're on a journey to learn excellence. I could happily merge this as-is, but here are some suggestions to fix. Keep in mind, anytime me, or someone else, suggests something you don't agree with, speak up! Often maintainers make mistakes. Of course do so carefully since it's often their project, but if they miss something that you didn't, let them know why.

purpleidea commented:
Review done!

lourencovales commented:
@purpleidea should be a-ok now

purpleidea commented:
Looks perfect, except can we put it into the net package instead of util please? Rebase to master too please and then I think it's merge time!

lourencovales commented:
Please review @purpleidea :-)

purpleidea commented:
merged \o/

woo! Thank you =D
